File: gtci/17-modified-binary-search/cpp/5-search-in-a-sorted-infinite-array.cpp
#include <cassert>
#include <climits>
#include <iostream>
#include <limits>
#include <vector>

class ArrayReader {
public:
  std::vector<int> arr;

  ArrayReader(const std::vector<int> &arr) { this->arr = arr; }

  virtual int get(int index) {
    if (index >= arr.size()) {
      return std::numeric_limits<int>::max();
    }
    return arr[index];
  }
};

class Solution {
public:
  static int searchInfiniteSortedArray(ArrayReader reader, int key) {
    int left = 0;
    int right = 1;

    // find a finite sized window to do the search
    while (reader.get(right) < key) {
      left = right;
      right *= 2;
    }

    // peform the search
    while (left <= right) {
      const int mid = left + (right - left) / 2;

      if (reader.get(mid) == key)
        return mid;
      else if (reader.get(mid) < key)
        left = mid + 1;
      else if (reader.get(mid) > key)
        right = mid - 1;
    }

    return -1;
  }
};

int main() {
    Solution s;

    ArrayReader reader1({4, 6, 8, 10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30});
    int key1 = 16;
    assert(s.searchInfiniteSortedArray(reader1, key1) == 6);

    ArrayReader reader2({1, 3, 8, 10, 15});
    int key2 = 15;
    assert(s.searchInfiniteSortedArray(reader2, key2) == 4);

    ArrayReader reader3({1, 3, 8, 10, 15});
    int key3 = 200;
    assert(s.searchInfiniteSortedArray(reader3, key3) == -1);

    std::cout << "All test cases passed." << std::endl;

    return 0;
}

File: gtci/17-modified-binary-search/python3/5-search-in-a-sorted-infinite-array.py
import math


class ArrayReader:
    def __init__(self, arr):
        self.arr = arr

    def get(self, index):
        if index >= len(self.arr):
            return math.inf
        return self.arr[index]


class Solution:
    def searchInfiniteSortedArray(self, reader, key):
        left = 0
        right = 1

        # find a finite window to perform the search in
        while reader.get(right) < key:
            left = right
            right *= 2  # double the window size on each step

        # perform the search
        while left <= right:
            mid = (left + right) // 2

            if reader.get(mid) == key:
                return mid
            elif reader.get(mid) < key:
                left = mid + 1
            elif reader.get(mid) > key:
                right = mid - 1

        return -1


if __name__ == "__main__":
    sol = Solution()

    reader1 = ArrayReader([4, 6, 10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30])
    assert sol.searchInfiniteSortedArray(reader1, 16) == 5
    assert sol.searchInfiniteSortedArray(reader1, 11) == -1

    reader2 = ArrayReader([1, 3, 8, 10, 15])
    assert sol.searchInfiniteSortedArray(reader2, 15) == 4
    assert sol.searchInfiniteSortedArray(reader2, 200) == -1

    print("All test cases passed.")
